src/raftstore/coprocessor/split_check/table.rs

```diff
@@ -0,0 +1,519 @@
+        }
+    }
+}
+
+/// Do some quick checks, true for skipping `check_key`.
+fn before_check(status: &mut TableStatus, engine: &DB, region: &Region) -> bool {
+    let (actual_start_key, actual_end_key) = match bound_keys(engine, region) {
+        Ok(Some((actual_start_key, actual_end_key))) => (actual_start_key, actual_end_key),
+        Ok(None) => return true,
+        Err(err) => {
+            error!(
+                "[region {}] failed to get region bound: {}",
+                region.get_id(),
+                err
+            );
+            return true;
+        }
+    };
+
+    if !keys::validate_data_key(&actual_start_key) || !keys::validate_data_key(&actual_end_key) {
+        return true;
+    }
+
+    let encoded_actual_start_key = keys::origin_key(&actual_start_key);
+    let encoded_actual_start_table_prefix =
+        if let Some(key) = extract_encoded_table_prefix(encoded_actual_start_key) {
+            key
+        } else {
+            return true;
+        };
+
+    let encoded_actual_end_key = keys::origin_key(&actual_end_key);
+    let encoded_actual_end_table_prefix =
+        if let Some(key) = extract_encoded_table_prefix(encoded_actual_end_key) {
+            key
+        } else {
+            return true;
+        };
+
+    // Table data starts with `TABLE_PREFIX`.
+    // Find out the actually range of this region by comparing with `TABLE_PREFIX`.
+    match (
+        encoded_actual_start_table_prefix[..table_codec::TABLE_PREFIX_LEN]
+            .cmp(table_codec::TABLE_PREFIX),
+        encoded_actual_end_table_prefix[..table_codec::TABLE_PREFIX_LEN]
+            .cmp(table_codec::TABLE_PREFIX),
+    ) {
+        // The range does not cover table data.
+        (Ordering::Less, Ordering::Less) | (Ordering::Greater, Ordering::Greater) => true,
+
+        // Following arms matches when the region contains table data.
+        // Covers all table data.
+        (Ordering::Less, Ordering::Greater) => false,
+        // The later part contains table data.
+        (Ordering::Less, Ordering::Equal) => {
+            // It starts from non-table area to table area,
+            // `encoded_actual_end_table_prefix` can be a split key, save it in context.
+            status.last_encoded_table_prefix = Some(encoded_actual_end_table_prefix.to_vec());
+            false
+        }
+        // Region is in table area.
+        (Ordering::Equal, Ordering::Equal) => {
+            if encoded_actual_start_table_prefix == encoded_actual_end_table_prefix {
+                // Same table.
+                true
+            } else {
+                // Different tables, choose `encoded_actual_end_table_prefix` as a split key.
+                // Note that table id does not grow by 1, so have to use
+                // `encoded_actual_end_table_prefix`.
+                // See more: https://github.com/pingcap/tidb/issues/4727
+                status.last_encoded_table_prefix = Some(encoded_actual_end_table_prefix.to_vec());
```

Comment:
Now PD uses start key to split the region, so here we will still use the start key.